src/OrchardCore/OrchardCore.Sms.Core/Services/TwilioSmsProvider.cs

```diff
@@ -79,14 +83,14 @@ public async Task<SmsResult> SendAsync(SmsMessage message)
         }
     }
 
-    private async Task EnsureClientIsIsInitializedAsync()
+    private async Task EnsureClientIsInitializedAsync()
     {
         if (_plainAuthToken == null)
         {
             var settings = (await _siteService.GetSiteSettingsAsync()).As<TwilioSettings>();
 
             _phoneNumber = settings.PhoneNumber;
-            var protector = _dataProtectionProvider.CreateProtector(SmsConstants.TwilioServiceName);
+            var protector = _dataProtectionProvider.CreateProtector(ProtectorName);
             _plainAuthToken = protector.Unprotect(settings.AuthToken);
 
             TwilioClient.Init(settings.AccountSID, _plainAuthToken);
```

Comment:
If there is no problem to call `TwilioClient.Init()` concurently, at least I would define a `TwilioSettings _settings` field that I would check and init after the initialization, this to add a minimum of atomicity. Then you can use this `_settings` field elsewhere, to access the phone number for example.

    private TwilioSettings _settings;

    private async Task EnsureClientIsIsInitializedAsync()
    {
        if (_settings is not null)
        {
            return;
        }

        var settings = (await _siteService.GetSiteSettingsAsync()).As<TwilioSettings>();
        var protector = _dataProtectionProvider.CreateProtector(SmsConstants.TwilioServiceName);

        var plainAuthToken = protector.Unprotect(settings.AuthToken);
        TwilioClient.Init(settings.AccountSID, plainAuthToken);

        _settings = settings;
    }



Otherwise use for example a `SemaphoreSlim` as we do in many places.

Reply:
I don't think there is an issue of using the `TwilioClient.Init()` all it does just sets the username/password internally. I am not sure why we need to save the settings in a private variable here. Not sure I understand the benefit or reasoning

Reply:
When you set multiple fields concurrently, if we don't want to use a lock because other codes are thread safe, one way to be atomic is to have a local object of a class having all these fields, then in the concurrent code create a new instance of this class, init all fields, and at the very end update this local object in one pass (this will only update the object reference pointer).

Otherwise you may ends up with unrelated field values, one important thing is to set what you are checking at the very end. Here I used `TwilioSettings` but you can use a dedicated class having all the fields that you need after in the code.

This is the principle, that said I saw that you then only use `PhoneNumber`, not `PlainAuthToken`, so we could only set and check a `PhoneNumber`, but in the future you may need other fields.

Reply:
Made some more changes